core/src/main/java/org/apache/iceberg/encryption/AesGcmOutputStream.java

```diff
@@ -0,0 +1,173 @@
+      throw new IOException(e);
+    }
+    this.random = new SecureRandom();
+    this.nonce = new byte[GCM_NONCE_LENGTH];
+    this.key = new SecretKeySpec(aesKey, "AES");
+    this.plaintextBlockBuffer = new byte[blockSize];
+    this.positionInBuffer = 0;
+    this.streamPosition = 0;
+    this.currentBlockIndex = 0;
+    this.fileAadPrefix = fileAadPrefix;
+
+    byte[] prefixBytes = ByteBuffer.allocate(PREFIX_LENGTH).order(ByteOrder.LITTLE_ENDIAN)
+        .put(MAGIC_ARRAY)
+        .putInt(blockSize)
+        .array();
+    targetStream.write(prefixBytes);
+  }
+
+  @Override
+  public void write(int b) throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void write(byte[] b)  throws IOException {
+    write(b, 0, b.length);
+  }
+
+  @Override
+  public void write(byte[] b, int off, int len) throws IOException {
+    int remaining = len;
+    int offset = off;
+
+    while (remaining > 0) {
+      int freeBlockBytes = blockSize - positionInBuffer;
+      int toWrite = freeBlockBytes <= remaining ? freeBlockBytes : remaining;
+
+      System.arraycopy(b, offset, plaintextBlockBuffer, positionInBuffer, toWrite);
+      positionInBuffer += toWrite;
+      if (positionInBuffer == blockSize) {
+        encryptAndWriteBlock();
+        positionInBuffer = 0;
+      }
+      offset += toWrite;
+      remaining -= toWrite;
+    }
+
+    streamPosition += len;
+  }
+
+  @Override
+  public long getPos() throws IOException {
+    return streamPosition;
+  }
+
+  @Override
+  public void flush() throws IOException {
+    targetStream.flush();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (positionInBuffer > 0) {
+      encryptAndWriteBlock();
+    }
+    targetStream.close();
+  }
+
+  private void encryptAndWriteBlock() throws IOException {
+    random.nextBytes(nonce);
+    GCMParameterSpec spec = new GCMParameterSpec(GCM_TAG_LENGTH_BITS, nonce);
+    try {
+      gcmCipher.init(Cipher.ENCRYPT_MODE, key, spec);
+    } catch (InvalidKeyException | InvalidAlgorithmParameterException e) {
+      throw new IOException("Failed to init GCM cipher", e);
+    }
+
+    byte[] aad = calculateAAD(fileAadPrefix, currentBlockIndex);
+    gcmCipher.updateAAD(aad);
+
+    byte[] cipherText = new byte[GCM_NONCE_LENGTH + positionInBuffer + GCM_TAG_LENGTH];
+    System.arraycopy(nonce, 0, cipherText, 0, GCM_NONCE_LENGTH);
+    try {
+      int encrypted = gcmCipher.doFinal(plaintextBlockBuffer, 0, positionInBuffer, cipherText, GCM_NONCE_LENGTH);
+      Preconditions.checkArgument((encrypted == (positionInBuffer + GCM_TAG_LENGTH)),
+          "Wrong length of encrypted output: " + encrypted + " vs " + (positionInBuffer + GCM_TAG_LENGTH));
+    } catch (GeneralSecurityException e) {
+      throw new IOException("Failed to encrypt", e);
+    }
+
+    currentBlockIndex++;
+
+    targetStream.write(cipherText);
+  }
+
+  static byte[] calculateAAD(byte[] fileAadPrefix, int currentBlockIndex) {
```

Comment:
This can go to a common place as well since both classes need it. We may have a util class, e.g. `GcmUtil`